content/non-blocking-communication-pt1.rst

```diff
@@ -1,33 +1,307 @@
+
+   The ``border data`` in dark blue has to be sent to the other
+   process, and the ``ghost data`` or ``halo data`` in dark green is
+   received from the other process.
+
+Inferior MPI workflows for stencil applications
+-----------------------------------------------
+
+A possible workflow for the code on these two processes looks like
+
+.. figure:: img/simplestencilworkflow.svg
+   :align: center
+
+   A simple parallel stencil-computation workflow. Depending on the
+   MPI implementation, doing the send first on both ranks can lead to
+   a deadlock.
+
+This may work sometimes, but it also might not if the buffers or
+number of messages gets large enough. Such code will not be
+portable.
+
+.. figure:: img/ring-stylestencilworkflow.svg
+   :align: center
+
+   A correct ring-style parallel stencil-computation workflow.
+
+This workflow will always work. This approach generalizes to larger
+rings, so long as alternating processes either send or respectively
+receive first. It also generalizes to multiple dimensions. However, it
+still requires synchronization between send and receive ranks. That
+synchronization is problematic when the time to send messages is
+comparable with either the amounts of compute work involved, or the
+variation in those amounts.
+
+Performant non-blocking MPI workflows for stencil applications
+--------------------------------------------------------------
+
+.. figure:: img/non-blocking-stylestencilworkflow.svg
+   :align: center
+
+   A correct non-blocking-style parallel stencil-computation
+   workflow. The MPI implementation may be able to do the
+   communication overlapped with the computation.
+
+This will also always work. It may be more efficient than the
+ring-style approach, but that depends on the implementation and
+environment, as well as the size of messages and amounts of compute
+that can participate in the overlapped region.
+
+The user is responsible for not using the memory area until after the
+message request has been waited upon.
+
+Non-blocking MPI send calls
+---------------------------
+
+An |term-MPI_Isend| is creates a send request and returns a request
+object. It may or may not have sent the message, or buffered it. The
+caller is responsible for not changing the buffer until after waiting
+upon the resulting request object.
+
+Call signature:
+
+.. code-block:: c
+
+    int MPI_Isend(const void* buf, int count, MPI_Datatype datatype,
+                  int dest, int tag, MPI_Comm comm, MPI_Request *request)
+
+.. note::
+
+    ``buf``, ``count`` and ``datatype`` describe the buffer to be sent
+    to ``dest`` rank of ``comm`` using tag ``tag``. The ``request`` object
+    that is returned must be used to wait on the communication later.
+
+Other calls exist for other sending modes, including buffered,
+synchronous, and ready-mode sends.
+
+
+Non-blocking MPI receive call
+-----------------------------
 
+An |term-MPI_Irecv| creates a receive request and returns a request
```

Comment:
How about: "An |term-MPI_Irecv| creates and returns a receive request in a `MPI_Request` object." To my non-native ear the current version sounds like the returned request is the return value.

Reply:
Agreed